### DISCUSSION TITLE
Use of WS as a separator in parsing sexpr

### DISCUSSION THREAD

**Discussion by caw** [lark-parser/lark#1271] [Q&A] [answered]
I have a simple sexpr parser:
```
start: value
value: list
      | string
      | number
list  : "(" [value (" " value)*] ")"
%import common.WS
%ignore WS
```

this fails on
```
(foo
(bar baz))
```
if there is no space ```" "``` after foo

I can fix this by setting the separator in the ```list``` rule to WS
```
start: value
value: list
      | string
      | number
list  : "(" [value (WS value)*] ")"
%import common.WS
%ignore WS
```

I get a correct parse, but is that the sensible way to do this? And I'm not sure how "%ignore WS" interacts with WS being in a rule.

with thanks

**Comment by MegaIng** [comment, marked as answer]
This is fine. The first example fails because you are explicitly requiring a space. Don't b so that it that is not the correct semantics. 

The second example will work, lark will prefer using a terminal in a rule instead of ignoring it.

For a slightly different syntax you can also have no seperator in the list at all.

**Reply by caw** [reply]
thank you - I appreciate your help.
the parse now returns a WS Token. Is there a way to prevent that being used in building the parse tree?

**Reply by caw** [reply]
Oh, and the suggestion to have no separator in the list is very interesting!

**Reply by MegaIng** [reply]
You can control how the Tree looks like, see Shaping the tree: https://lark-parser.readthedocs.io/en/latest/tree_construction.html#shaping-the-tree
